samples/bluetooth/peripheral_hids_mouse/src/main.c

```diff
@@ -460,66 +467,53 @@ static struct bt_conn_auth_cb conn_auth_callbacks;
 #endif
 
 
-void button_pressed(struct device *gpio_dev, struct gpio_callback *cb,
-		    u32_t pins)
+void button_changed(u32_t button_state, u32_t has_changed)
 {
+	bool data_to_send = false;
 	struct mouse_pos pos;
-	int err;
+	u32_t buttons = button_state & has_changed;
 
 	memset(&pos, 0, sizeof(struct mouse_pos));
 
-	if (pins & (1 << SW0_GPIO_PIN)) {
+	if (buttons & KEY_LEFT_MASK) {
 		pos.x_val -= MOVEMENT_SPEED;
 		printk("%s(): left\n", __func__);
+		data_to_send = true;
 	}
-	if (pins & (1 << SW1_GPIO_PIN)) {
+	if (buttons & KEY_UP_MASK) {
 		pos.y_val -= MOVEMENT_SPEED;
 		printk("%s(): up\n", __func__);
+		data_to_send = true;
 	}
-	if (pins & (1 << SW2_GPIO_PIN)) {
+	if (buttons & KEY_RIGHT_MASK) {
 		pos.x_val += MOVEMENT_SPEED;
 		printk("%s(): right\n", __func__);
+		data_to_send = true;
 	}
-	if (pins & (1 << SW3_GPIO_PIN)) {
+	if (buttons & KEY_DOWN_MASK) {
 		pos.y_val += MOVEMENT_SPEED;
 		printk("%s(): down\n", __func__);
+		data_to_send = true;
 	}
 
-	err = k_msgq_put(&hids_queue, &pos, K_NO_WAIT);
-
-	if (err) {
-		printk("No space in the queue for button pressed\n");
-		return;
-	}
+	if (data_to_send) {
```

Comment:
is this new variable really needed? could you use ```has_changed``` instead?

Reply:
Without this variable when button is released useless report is send with all values set to 0. I will update the name.

Reply:
The value of has_changed has ones on the positions of the buttons that has been changed. We could use buttons internal variable, but then, to be safe we should check only the bits that w are processing - this would require additional masking with all the buttons masks. It may look simpler - we are removing 5 lines, but it is more error prone - if we remove any button from here we have to update mask.